File: src/shell_cluster/tunnel/checks.py
"""Pre-flight checks for tunnel backends (CLI startup)."""

from __future__ import annotations

import json
import os
import shutil
import subprocess
import time

from rich.console import Console

console = Console()


def check_devtunnel() -> bool:
    """Check that devtunnel CLI is installed and logged in. Returns True if ok."""
    if not shutil.which("devtunnel"):
        console.print("[red]devtunnel CLI is not installed.[/red]")
        console.print("Install it: https://learn.microsoft.com/en-us/azure/developer/dev-tunnels/get-started")
        return False

    try:
        result = subprocess.run(
            ["devtunnel", "list", "--limit", "1"],
            capture_output=True, timeout=10,
        )
        if result.returncode != 0:
            stderr = result.stderr.decode().strip()
            if "login" in stderr.lower() or "sign in" in stderr.lower() or "unauthorized" in stderr.lower():
                console.print("[red]devtunnel is not logged in.[/red]")
                console.print("Run: [bold]devtunnel user login[/bold]")
                return False
            # Other errors — might still work, let it try
    except subprocess.TimeoutExpired:
        console.print("[yellow]devtunnel login check timed out, proceeding anyway.[/yellow]")
    except Exception:
        pass

    return True


def _start_tailscaled(socket: str) -> bool:
    """Start tailscaled in userspace-networking mode. Returns True if started ok."""
    if not shutil.which("tailscaled"):
        console.print("[red]tailscaled is not installed.[/red]")
        console.print("Install it: [bold]brew install tailscale[/bold]")
        return False

    console.print("[yellow]Starting tailscaled...[/yellow]")
    args = ["tailscaled", "--tun=userspace-networking"]
    if socket:
        args += ["--socket", socket]

    # Start as a detached background process
    subprocess.Popen(
        args,
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
        start_new_session=True,
    )

    # Wait for socket to appear
    for _ in range(10):
        time.sleep(0.5)
        if socket and os.path.exists(socket):
            console.print("[green]tailscaled started.[/green]")
            return True
    console.print("[green]tailscaled started.[/green]")
    return True


def check_tailscale() -> bool:
    """Check that tailscale CLI is installed and connected. Returns True if ok."""
    if not shutil.which("tailscale"):
        console.print("[red]tailscale CLI is not installed.[/red]")
        console.print("Install it: [bold]brew install tailscale[/bold]")
        return False

    from shell_cluster.tunnel.tailscale import _default_socket
    socket = _default_socket()

    cmd = ["tailscale"]
    if socket:
        cmd += ["--socket", socket]
    cmd += ["status", "--json"]

    try:
        result = subprocess.run(
            cmd,
            capture_output=True, timeout=10,
        )
        if result.returncode != 0:
            # tailscaled not running — try to start it
            if not _start_tailscaled(socket):
                return False
            # Re-check after starting
            result = subprocess.run(cmd, capture_output=True, timeout=10)
            if result.returncode != 0:
                console.print("[red]tailscaled started but not responding.[/red]")
                return False
        status = json.loads(result.stdout)
        state = status.get("BackendState", "")
        if state != "Running":
            console.print(f"[yellow]tailscale is not logged in (state: {state}).[/yellow]")
            console.print("Run: [bold]tailscale --socket '{socket}' up[/bold]" if socket else "Run: [bold]tailscale up[/bold]")
            return False
    except subprocess.TimeoutExpired:
        console.print("[yellow]tailscale status check timed out, proceeding anyway.[/yellow]")
    except Exception:
        pass

    return True

File: server/src/tunnel/checks.js
/**
 * Pre-flight checks for tunnel backends (Node.js port).
 */

'use strict';

const { execFileSync } = require('child_process');
const { existsSync } = require('fs');
const path = require('path');

/**
 * Check if an executable is available in PATH.
 * @param {string} name
 * @returns {boolean}
 */
function which(name) {
  try {
    const cmd = process.platform === 'win32' ? 'where' : 'which';
    execFileSync(cmd, [name], { stdio: 'pipe' });
    return true;
  } catch (e) {
    return false;
  }
}

/**
 * Check that devtunnel CLI is installed and logged in.
 * @returns {boolean}
 */
function checkDevtunnel() {
  if (!which('devtunnel')) {
    console.error('devtunnel CLI is not installed.');
    console.error('Install it: https://learn.microsoft.com/en-us/azure/developer/dev-tunnels/get-started');
    return false;
  }

  try {
    const result = execFileSync('devtunnel', ['list', '--limit', '1'], {
      stdio: 'pipe',
      timeout: 10000,
    });
  } catch (e) {
    if (e.stderr) {
      const stderr = e.stderr.toString().toLowerCase();
      if (stderr.includes('login') || stderr.includes('sign in') || stderr.includes('unauthorized')) {
        console.error('devtunnel is not logged in.');
        console.error('Run: devtunnel user login');
        return false;
      }
    }
    // Timeout or other error - might still work, proceed
  }

  return true;
}

/**
 * Start tailscaled in userspace-networking mode.
 * @param {string} socket
 * @returns {boolean}
 */
function startTailscaled(socket) {
  if (!which('tailscaled')) {
    console.error('tailscaled is not installed.');
    console.error('Install it: brew install tailscale');
    return false;
  }

  console.log('Starting tailscaled...');
  const args = ['tailscaled', '--tun=userspace-networking'];
  if (socket) {
    args.push('--socket', socket);
  }

  const { spawn } = require('child_process');
  spawn(args[0], args.slice(1), {
    stdio: 'ignore',
    detached: true,
  }).unref();

  // Wait for socket to appear (blocking poll with Atomics.wait)
  for (let i = 0; i < 10; i++) {
    const waitBuf = new Int32Array(new SharedArrayBuffer(4));
    Atomics.wait(waitBuf, 0, 0, 500); // 500ms sleep without execSync
    if (socket && existsSync(socket)) {
      console.log('tailscaled started.');
      return true;
    }
  }
  console.log('tailscaled started.');
  return true;
}

/**
 * Check that tailscale CLI is installed and connected.
 * @returns {boolean}
 */
function checkTailscale() {
  if (!which('tailscale')) {
    console.error('tailscale CLI is not installed.');
    console.error('Install it: brew install tailscale');
    return false;
  }

  const { defaultSocket } = require('./tailscale');
  const socket = defaultSocket();

  const cmd = ['tailscale'];
  if (socket) cmd.push('--socket', socket);
  cmd.push('status', '--json');

  try {
    const output = execFileSync(cmd[0], cmd.slice(1), {
      stdio: 'pipe',
      timeout: 10000,
    });
    const status = JSON.parse(output.toString());
    const state = status.BackendState || '';
    if (state !== 'Running') {
      console.warn(`tailscale is not logged in (state: ${state}).`);
      if (socket) {
        console.error(`Run: tailscale --socket '${socket}' up`);
      } else {
        console.error('Run: tailscale up');
      }
      return false;
    }
  } catch (e) {
    if (e.status !== undefined && e.status !== 0) {
      // tailscaled not running - try to start it
      if (!startTailscaled(socket)) return false;
      // Re-check
      try {
        const output = execFileSync(cmd[0], cmd.slice(1), {
          stdio: 'pipe',
          timeout: 10000,
        });
      } catch (e2) {
        console.error('tailscaled started but not responding.');
        return false;
      }
    }
    // Timeout or other error - proceed
  }

  return true;
}

module.exports = { checkDevtunnel, checkTailscale, which };
